drivers/counter/counter_nxp_mrt.c

```diff
@@ -0,0 +1,297 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_COUNTER_LOG_LEVEL);
+
+/* Specific for each channel */
+struct nxp_mrt_channel_data {
+	uint32_t top;
+	counter_top_callback_t cb;
+	void *user_data;
+};
+
+/* Shared between all channels */
+struct nxp_mrt_config {
+	struct counter_config_info info;
+	MRT_Type *base;
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+	void (*irq_config_func)(const struct device *dev);
+	struct nxp_mrt_channel_data *data;
+	const struct device *const *const channels;
+};
+
+static int nxp_mrt_stop(const struct device *dev)
+{
+	const struct nxp_mrt_config *config = dev->config;
+	struct nxp_mrt_channel_data *data = dev->data;
+	MRT_Type *base = config->base;
+	int channel_id = data - config->data;
+
+	LOG_DBG("MRT@%p channel %d stopped", base, channel_id);
+	LOG_WRN("MRT channel resets upon stopping");
+
+	/* LOAD bit and 0 ivalue allows us to forcibly stop the timer */
+	base->CHANNEL[channel_id].INTVAL = MRT_CHANNEL_INTVAL_LOAD(1);
+
+	return 0;
+}
+
+static int nxp_mrt_start(const struct device *dev)
+{
+	const struct nxp_mrt_config *config = dev->config;
+	struct nxp_mrt_channel_data *data = dev->data;
+	MRT_Type *base = config->base;
+	int channel_id = data - config->data;
+
+	if (data->top <= 1) {
+		/* Zephyr API says default should be max top value */
+		LOG_INF("\"Started\" MRT@%p channel %d with default value %d",
+				base, channel_id, config->info.max_top_value);
+		data->top = config->info.max_top_value;
+	}
+
+	/* Start with previously configured top value (if already running this has no effect) */
+	base->CHANNEL[channel_id].INTVAL = data->top;
+
+	LOG_DBG("MRT@%p channel %d started with top value %d", base, channel_id, data->top);
+
+	return 0;
+}
+
+static int nxp_mrt_get_value(const struct device *dev, uint32_t *ticks)
+{
+	const struct nxp_mrt_config *config = dev->config;
+	struct nxp_mrt_channel_data *data = dev->data;
+	MRT_Type *base = config->base;
+	int channel_id = data - config->data;
+
+	*ticks = base->CHANNEL[channel_id].TIMER & MRT_CHANNEL_TIMER_VALUE_MASK;
+
+	return 0;
+}
+
+
+static int nxp_mrt_set_top_value(const struct device *dev, const struct counter_top_cfg *cfg)
+{
+	const struct nxp_mrt_config *config = dev->config;
+	struct nxp_mrt_channel_data *data = dev->data;
+	MRT_Type *base = config->base;
+	/* By default in Zephyr API, the counter resets on changing top value */
+	bool reset = !(cfg->flags & COUNTER_TOP_CFG_DONT_RESET);
+	int channel_id = data - config->data;
+	bool active = base->CHANNEL[channel_id].STAT & MRT_CHANNEL_STAT_RUN_MASK;
+	uint32_t current_val = base->CHANNEL[channel_id].TIMER & MRT_CHANNEL_TIMER_VALUE_MASK;
+	int ret = 0;
+
+	/* Store for use by counter_start */
+	data->top = cfg->ticks;
+
+	/* Used by ISR */
+	data->cb = cfg->callback;
+	data->user_data = cfg->user_data;
+
+
+	/* If not currently running, nothing else to do but wait for counter_start */
+	if (!active) {
```

Comment:
I don't follow this. Why cannot we change the top value if counter is not active?

Reply:
setting the value will start the timer

Reply:
I improved this comment on line 107